### ISSUE TITLE
Nested ContainerBuilder created with BeginLifetimeScope shares BuildCallback list with its parent ContainerBuilder

### ISSUE THREAD

**Issue by evil-shrike** [autofac/Autofac#912]
```csharp
var builder = new ContainerBuilder();
builder.RegisterModule(new MyModule ());
var container = builder.Build(); // here I'm getting InvalidOperationException
```

Inside MyModule I create a nested scope and register a BuildCallback for the **nested** container builder:
```csharp
    public class MyModule : Module
    {
        protected override void Load(ContainerBuilder containerBuilder)
        {
            containerBuilder.RegisterBuildCallback(container =>
            {
                m_appScope = container.BeginLifetimeScope(nestedBuilder =>
                {
                    nestedBuilder.RegisterBuildCallback((c) => {
                      //
                    });
                });
                m_app = m_appScope.Resolve<IMyService>();
            });
        }
  }
```
But it seems that these two builders (the outer and the nested) share the same callbacks list. As I'm gettting an exception "Collection was modified; enumeration operation may not execute" on `Build()` call for the outer builder:

```
System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
Unhandled Exception:   at System.ThrowHelper.ThrowInvalidOperationException_InvalidOpera tion_EnumFailedVersion()
   at System.Collections.Generic.List`1.Enumerator.MoveNextRare()
   at Autofac.ContainerBuilder.Build(ContainerBuildOptions options)
```

Autofac 2.8.0

**Comment by tillig** [comment]
We can definitely look into this but from a general standpoint I'd recommend separating your _app startup_ code - the creation of an application level scope and the execution of your app factory stuff - from the population and building of the _container_. I can see how it might be easy to try tying it all together but I can also see high potential for getting into challenging and complicated situations like this.

Like I said, though, we will look into it. Thanks for the report!

**Comment by alistairjevans** [comment]
This appears to have been fixed by the changes in the #1054 PR, as part of fixing #985.

Separate PR incoming with a proving test from the originally reported behaviour.

**Comment by tillig** [comment]
PR accepted so we won't regress this.
